lib/tsconfig-parser.js

```diff
@@ -0,0 +1,144 @@
+'use strict';
+
+const path = require('path');
+const fs = require('fs');
+const ts = require('typescript');
+
+//------------------------------------------------------------------------------
+// Environment calculation
+//------------------------------------------------------------------------------
+
+/**
+ * Maps tsconfig paths to their corresponding file contents and resulting watches
+ * @type {Map<string, ts.WatchOfConfigFile<ts.SemanticDiagnosticsBuilderProgram>>}
+ */
+const knownWatchProgramMap = new Map();
+
+/**
+ * Maps file paths to their set of corresponding watch callbacks
+ * There may be more than one per file if a file is shared between projects
+ * @type {Map<string, ts.FileWatcherCallback>}
+ */
+const watchCallbackTrackingMap = new Map();
+
+/**
+ * Holds information about the file currently being linted
+ * @type {{code: string, filePath: string}}
+ */
+const currentLintOperationState = {
+  code: '',
+  filePath: ''
+};
+
+/**
+ * Appropriately report issues found when reading a config file
+ * @param {ts.Diagnostic} diagnostic The diagnostic raised when creating a program
+ * @returns {void}
+ */
+function diagnosticReporter(diagnostic) {
+  throw new Error(
+    ts.flattenDiagnosticMessageText(diagnostic.messageText, ts.sys.newLine)
+  );
+}
+
+const noopFileWatcher = { close: () => {} };
+
+/**
+ * Calculate project environments using options provided by eslint and paths from config
+ * @param {string} code The code being linted
+ * @param {Object} options Options provided by ESLint core
+ * @param {string} options.cwd The current working directory for the eslint process
+ * @param {string} options.filePath The path of the file being parsed
+ * @param {Object} extra Validated parser options
```

Comment:
This object is only used for a single property, just pass the property?

Reply:
Fixed.